ssl/ssl_lib.c

```diff
@@ -7211,3 +7260,208 @@ int SSL_stream_conclude(SSL *ssl, uint64_t flags)
     return 0;
 #endif
 }
+
+int SSL_add_expected_rpk(SSL *s, EVP_PKEY *rpk, int mtype)
+{
+    unsigned char *i2dbuf = NULL;
+    SSL_DANE *dane = SSL_get0_dane(s);
+    size_t i2dlen;
+    unsigned char dgstbuf[EVP_MAX_MD_SIZE];
+    unsigned int dgstlen;
+    unsigned char *data;
+    size_t len;
+    int ret;
+    const EVP_MD *md = NULL;
+    int ok = 0;
+
+    if (dane == NULL)
+        return 0;
+
+    switch (mtype) {
+    case DANETLS_MATCHING_FULL:
+    case DANETLS_MATCHING_2256:
+    case DANETLS_MATCHING_2512:
+        break;
+    default:
+        goto err;
+    }
```

Comment:
```suggestion
    if (mtype > dane->mdmax)
        goto err;
```
This allows users to make use of novel and private matching type code points registered via `SSL_CTX_dane_mtype_set()` or later added to OpenSSL itself without changing the RPK code. The data type of `mtype` should be `uint8_t`.